MetaMask/metamask-extension
# Swaps: Add conditional routing to new APIs based on a feature flag

## Explanation
This PR adds conditional routing to either v1 or v2 APIs for Swaps, based on feature flags from our backend. The goal is to migrate to v2 APIs only, but we want to have a fallback solution to v1, just to be on the safe side.

## Manual testing steps
v1 API calls:
- Open the `swaps.util.js` file and return the following from the `fetchSwapsFeatureFlags` function:
```
return {
    bsc: {
      mobile_active: false,
      extension_active: false,
      fallback_to_v1: true,
    },
    ethereum: {
      mobile_active: false,
      extension_active: false,
      fallback_to_v1: true,
    },
    polygon: {
      mobile_active: false,
      extension_active: false,
      fallback_to_v1: false,
    },
  };
```
- Run `yarn start` with a branch that includes this PR
- Open the extension
- On the main page click on Swaps
- Check that network calls are going to our old APIs

v2 API calls:
- Open the `swaps.util.js` file and return the following from the `fetchSwapsFeatureFlags` function:
```
return {
    bsc: {
      mobile_active: false,
      extension_active: true,
      fallback_to_v1: true,
    },
    ethereum: {
      mobile_active: false,
      extension_active: true,
      fallback_to_v1: true,
    },
    polygon: {
      mobile_active: false,
      extension_active: true,
      fallback_to_v1: false,
    },
  };
```
- Run `yarn start` with a branch that includes this PR
- Open the extension
- On the main page click on Swaps
- Check that network calls are going to our new APIs (`https://api2.metaswap.codefi.network/`)

## Screenshots
v2 /featureFlags API:
![image](https://user-images.githubusercontent.com/80175477/124759089-8e5f4f80-df2f-11eb-98f0-e5684a2ba4aa.png)

v1 API calls:
![image](https://user-images.githubusercontent.com/80175477/124758834-44766980-df2f-11eb-9fe0-d6f3e759bc48.png)

v1 successful swapping:
![image](https://user-images.githubusercontent.com/80175477/124761155-c9628280-df31-11eb-9f37-10b18960fffc.png)

v2 API calls:
![image](https://user-images.githubusercontent.com/80175477/124759141-9dde9880-df2f-11eb-881a-94eb7b78ba67.png)
